### ISSUE TITLE
Numeric helper methods: Gigabytes (GB) vs Gibibytes (GiB)

### ISSUE THREAD

**Issue by julienbourdeau** [rails/rails#40054]
## TL;DR

I'm looking for solution to update the byte size methods to support both decimal and binary. Ideally without breaking backward compatibility.

Please share your thoughts 🙏 

## Context

I work on the Algolia dashboard, a Rails 5.2 application.

We recently discovered a bug where we were displaying a wrong file size. We'd pull the data size in bytes from another API, convert it to gigabytes and found that the total size was off by a few gigabytes.

This is how I discovered the difference between Gigabytes GB and Gibibytes (GiB).

## Gigabytes (GB) vs Gibibytes (GiB)

I have discovered that the term Gigabytes is supposed to be used in decimal (base 10) while Gibibytes should be used to describe size in binary (base 2). 

* 1 GB = 1000³
* 1 Gib = 1024³

I realized that many people use the word gigabytes for size in binary. I have done it for years, I had no idea it was wrong.

> In 1998 the International Electrotechnical Commission (IEC) published standards for binary prefixes, requiring that the gigabyte strictly denote 10003 bytes and gibibyte denote 10243 bytes. By the end of 2007, the IEC Standard had been adopted by the IEEE, EU, and NIST, and in 2009 it was incorporated in the International System of Quantities. Nevertheless, the term gigabyte continues to be widely used with the two different meanings.
-- [Wikipedia](https://en.wikipedia.org/wiki/Gigabyte)

<img width="352" alt="Screenshot 2020-08-15 at 09 17 34" src="https://user-images.githubusercontent.com/1525636/90308514-5d71c880-dee0-11ea-91b2-9628a40446d5.png">

## Update numeric methods

Rails provide helpers to get a file size easily like `4.kilobytes` or `1.gigabyte`. We could change this methods to return results in base 10 and introduce new `4.kibibytes` and `1.gibibytes`.

This is the first solution that comes to mind, but it has big drawbacks.

| Method     | Size                  | Method     | Size                  |
|------------|-----------------------|------------|-----------------------|
| 1.kilobyte | 1_000                 | 1.kibibyte | 1_024                 |
| 1.megabyte | 1_000_000             | 1.mebibyte | 1_048_576             |
| 1.gigabyte | 1_000_000_000         | 1.gibibyte | 1_073_741_824         |
| 1.terabyte | 1_000_000_000_000     | 1.tebibyte | 1_099_511_627_776     |
| 1.petabyte | 1_000_000_000_000_000 | 1.pebibyte | 1_125_899_906_842_624 |

### Reasons to do it

1. I believe this would avoid bugs and **help people discovering the difference between the two**.
2. Dividing by `1_000_000_000` in your code is not as clear as `1.gigabyte`. Expressiveness is what make Rails and Ruby such a joy to use.

### Reasons not to

1. It's a massive breaking change, all gems relying on those methods will start failing. Maintaining backward compatibility with earlier version will be **painful**.
2. Main providers like [AWS S3 uses binary](https://forums.aws.amazon.com/thread.jspa?threadID=151161)
2. Nobody cares about the difference 
 
 
## Alternative solutions

For backward compatibility, we added not methods in our apps prefixed with `actual_`.

* 1.kilobytes = 1024
* 1.actual_kilobytes = 1000

Another fix I can think of is to introduce methods like `millions`,  `thousands` and such so we can write `1.million.bytes`.

Going a big crazy here, we could also add simple `giga`, `mega`, or `kilo` to be able to write `2.mega.bytes` vs `2.megabutes`. I'm afraid this would be error-prone and confusing.

**Comment by terracatta** [comment]
Discussed in https://github.com/rails/rails/issues/33130 and https://github.com/rails/rails/pull/7835, I don't think the appetite for this has changed in the months/years since.

Basically you could once pass a prefix to change the base from 1024 to 1000, this capability was depreciated in Rails 5. 

I write an app that needs to display hard drive sizes so I simply did the following. 

```ruby
module ActiveSupport
  module NumberHelper
    class NumberToHumanSizeConverter < NumberConverter
    private
      # We need this monkey patch because Rails has dropped support for "si"
      # prefixed sizes. Hard drive manufacturers calculate bytes -> to kb by
      # dividing by 1000 not 1024. In order for us to use this helper in that
      # context we need the ability to change this base.
      def base
        options[:base] || 1024
      end
    end
  end
end
```

There are also gems like https://github.com/olbrich/ruby-units which you can use to get the more "correct" names if you choose.

**Comment by julienbourdeau** [comment]
I would definitely prefer to introduce new methods and keep the existing behavior.

It's very important that we can predict what `3.gigabytes` is going to return. Changing the existing method will likely create issues where gems expect 1024 and get 1000.

**Comment by KapilSachdev** [comment]
@julienbourdeau  I think we can close this issue, as @terracatta pointed out and discussions in #33130 sums up that this issue is not going to be worked upon any further.

**Comment by julienbourdeau** [comment]
I understand 👍

**Comment by OlegChuev** [comment]
For anyone looking to use the built-in `number_to_human_size` method without any additional gems, here’s a monkey-patched version:

```ruby
# config/initializers/number_to_human_size_converter.rb
# frozen_string_literal: true

# Monkey patched the NumberConverter class to support tracking storage usage in ISO metrics as well as SI metrics.
# Added a new option, :prefix.
# When options[:prefix] is set to 'iec', calculations use a 1000 base. Otherwise, they default to using a 1024 base.
#
# Examples:
#   - number_to_human_size(1000, prefix: :iec) => 1 KiB
#   - number_to_human_size(1000) => 1000 Bytes
#
# https://github.com/rails/rails/issues/40054
#
#
# NOTE: Rails dropped initial support of the :prefix option, so we have to add it manually:
# https://github.com/rails/rails/pull/7835/files
# https://github.com/rails/rails/pull/21191
#
module ActiveSupport
  module NumberHelper
    # https://github.com/rails/rails/blob/main/activesupport/lib/active_support/number_helper/number_converter.rb
    class NumberConverter
      DEFAULTS[:human][:storage_units][:iec_units] = {
        byte: "Bytes",
        kb: "KiB",
        mb: "MiB",
        gb: "GiB",
        tb: "TiB",
        pb: "PiB",
        eb: "EiB",
        zb: "ZiB"
      }
    end

    # https://github.com/rails/rails/blob/main/activesupport/lib/active_support/number_helper/number_to_human_size_converter.rb
    class NumberToHumanSizeConverter < NumberConverter
      private

      def storage_unit_key
        key_end = smaller_than_base? ? "byte" : STORAGE_UNITS[exponent]

        "human.storage_units.#{custom_units}.#{key_end}"
      end

      def base
        case options[:prefix]
        when :iec
          1000
        else
          1024
        end
      end

      def custom_units
        case options[:prefix]
        when :iec
          "iec_units"
        else
          "units"
        end
      end
    end
  end
end
```

**Comment by bdorry** [comment]
@OlegChuev I'm looking at your monkey-patch and are the base options not inverted? The default option would be base of 1024 and the examples in the comments return:

`irb(main):002:0> ActiveSupport::NumberHelper.number_to_human_size(1024)
=> "1 KB"
irb(main):003:0> ActiveSupport::NumberHelper.number_to_human_size(1024, prefix: :iec)
=> "1.02 KiB"`

**Comment by OlegChuev** [comment]
@bdorry Ah, my bad! Sorry about that. It should’ve been reverted, you’re right. I’ve updated my solution above 🙂
